ios/browser/api/brave_wallet/brave_wallet_api.mm

```diff
@@ -0,0 +1,185 @@
+#include "brave/ios/browser/api/brave_wallet/brave_wallet.mojom.objc+private.h"
+#include "brave/ios/browser/api/brave_wallet/brave_wallet_provider_delegate_ios+private.h"
+#include "brave/ios/browser/api/brave_wallet/brave_wallet_provider_delegate_ios.h"
+#include "brave/ios/browser/brave_wallet/brave_wallet_service_factory.h"
+#include "brave/ios/browser/brave_wallet/json_rpc_service_factory.h"
+#include "brave/ios/browser/brave_wallet/keyring_service_factory.h"
+#include "brave/ios/browser/brave_wallet/tx_service_factory.h"
+#include "ios/chrome/browser/browser_state/chrome_browser_state.h"
+#include "ios/chrome/browser/browser_state/chrome_browser_state_manager.h"
+#include "ios/chrome/browser/content_settings/host_content_settings_map_factory.h"
+#include "ui/base/resource/resource_bundle.h"
+
+BraveWalletProviderScriptKey const BraveWalletProviderScriptKeyEthereum =
+    @"ethereum_provider.js";
+BraveWalletProviderScriptKey const BraveWalletProviderScriptKeySolana =
+    @"solana_provider.js";
+BraveWalletProviderScriptKey const BraveWalletProviderScriptKeySolanaInternal =
+    @"solana_provider_internal.js";
+
+#if !defined(__has_feature) || !__has_feature(objc_arc)
+#error "This file requires ARC support."
+#endif
+
+@implementation BraveWalletAPI {
+  ChromeBrowserState* _mainBrowserState;  // NOT OWNED
+  NSMutableDictionary<NSNumber* /* BraveWalletCoinType */,
+                      NSDictionary<BraveWalletProviderScriptKey, NSString*>*>*
+      _providerScripts;
+}
+
+- (instancetype)initWithBrowserState:(ChromeBrowserState*)mainBrowserState {
+  if ((self = [super init])) {
+    _mainBrowserState = mainBrowserState;
+    _providerScripts = [[NSMutableDictionary alloc] init];
+  }
+  return self;
+}
+
++ (id<BraveWalletBlockchainRegistry>)blockchainRegistry {
+  auto* registry = brave_wallet::BlockchainRegistry::GetInstance();
+  return [[BraveWalletBlockchainRegistryImpl alloc]
+      initWithBlockchainRegistry:registry];
+}
+
+- (nullable id<BraveWalletEthereumProvider>)
+    ethereumProviderWithDelegate:(id<BraveWalletProviderDelegate>)delegate
+               isPrivateBrowsing:(bool)isPrivateBrowsing {
+  auto* browserState = _mainBrowserState;
+  if (isPrivateBrowsing) {
+    browserState = browserState->GetOffTheRecordChromeBrowserState();
+  }
+
+  auto* json_rpc_service =
+      brave_wallet::JsonRpcServiceFactory::GetServiceForState(browserState);
+  if (!json_rpc_service) {
+    return nil;
+  }
+
+  auto* tx_service =
+      brave_wallet::TxServiceFactory::GetServiceForState(browserState);
+  if (!tx_service) {
+    return nil;
+  }
+
+  auto* keyring_service =
+      brave_wallet::KeyringServiceFactory::GetServiceForState(browserState);
+  if (!keyring_service) {
+    return nil;
+  }
+
+  auto* brave_wallet_service =
+      brave_wallet::BraveWalletServiceFactory::GetServiceForState(browserState);
+  if (!brave_wallet_service) {
+    return nil;
+  }
+
+  auto* provider = new brave_wallet::EthereumProviderImpl(
```

Comment:
This leaks.
The `brave_wallet::EthereumProviderImpl` takes a pointer, stores it as a `raw_ptr` but it never deletes it for you

Reply:
Fixed in https://github.com/brave/brave-core/pull/14703/commits/880af8f9b696d95a7013ac4e8e37ced6089b1e4c